### PR TITLE
Bidi lstm

### PR DESCRIPTION
**Pull request by kali**

### CODE REVIEW

**Comment by kali** [comment]
@bminixhofer this is only preliminary work. but bidi should be easy now :)

**Comment by kali** [comment]
@bminixhofer made some progress on that front. I wanted to test against bidaf, but bidaf has dynamicly sized tensors. I'm keeping that nightmare for later.

So would have a network with bidi lstm that I could use to test against ? Bonus points if you have input examples and matching outputs :)

**Comment by kali** [comment]
@bminixhofer You can also try the branch, it may actually work :)

**Comment by bminixhofer** [comment]
Thanks a lot!

I didn't have a model ready but I'm training one now at https://app.wandb.ai/bminixhofer/nnsplit/runs/1zw7nmkj (byte-level sentence boundary detection on German text).

I'll try it when its done and I can send you a minimal example if it doesn't work.

**Comment by bminixhofer** [comment]
I've made a small example. You can download the model from [here](
https://app.wandb.ai/bminixhofer/nnsplit/runs/4phefspj/files/model).
The code (including exporting to ONNX) is in [this file](https://app.wandb.ai/bminixhofer/nnsplit/runs/4phefspj/files/code/train/models.py).

Ideally this would work using tract instead of `onnxruntime`:

```python
import onnxruntime
import numpy as np

# list of bytes, sentence boundary at byte 16
x = list("Das ist ein Test Das ist noch ein Test.".encode("utf-8"))
x = np.array(x, dtype=np.uint8)[None, :]  # shape 1 x len

sess = onnxruntime.InferenceSession("model.onnx")
input_name = sess.get_inputs()[0].name

# shape 1 x len x 2
# value 0 in last dimension are token boundaries, value 1 are sentence boundaries
# outputs are logits, in practice there would be a sigmoid afterwards
out = sess.run(None, {input_name: x})[0]

# should be 16
print(np.argmax(out[0, :, 1]))
```

The model has dynamic axes at dimension zero (batch) and one (length) of the input, not sure if you were referring to that by dynamically sized tensors. If that is out of scope, I can convert it to fixed size dimensions.

**Comment by kali** [comment]
Thanks, @bminixhofer, I will have a look, I think it's exactly what I need to move forward.

**Comment by bminixhofer** [comment]
Great! Let me know if you need any changes. I'm not sure how practical the model is since I'm using dynamic axes and `batch_first=True` in PyTorch (the graph looks pretty complicated because of that).

**Comment by bminixhofer** [comment]
I just fixed a small mistake in the example (I forgot padding, and I mixed up the token and sentence indices). Shouldn't make much of a difference, but now it's correct.

**Comment by kali** [comment]
@bminixhofer All right, there was a pseudo-bug in the QLinearConv that I hope I have fixed (will push in this branch soon). Basically the spec does not say what type the bias should be (i32 or i8 as the input), and the analyse code was a bit eager to choose a side. (The evaluation was more flexible as it is shared between all variants of convolution, it just assumes the bias vector can be cast to the convolution accumulator type).

**Comment by kali** [comment]
I had to tweak a bit the input shape, in a way that I'm not 100% is consistent with what you were expecting, btw.

```
TSAR 06/06 11:17 ~/dev/snips/tract/cli% cargo run -- model.onnx -i 100x10xi8 --pass analyse dump                                                                                                                            bidi-lstmU
   Compiling tract-hir v0.6.4-pre (/home/kali/dev/snips/tract/hir)
   Compiling tract-kaldi v0.6.4-pre (/home/kali/dev/snips/tract/kaldi)
   Compiling tract-tensorflow v0.6.4-pre (/home/kali/dev/snips/tract/tensorflow)
   Compiling tract-onnx v0.6.4-pre (/home/kali/dev/snips/tract/onnx)
   Compiling tract v0.6.4-pre (/home/kali/dev/snips/tract/cli)
    Finished dev [unoptimized + debuginfo] target(s) in 7.30s
     Running `/home/kali/dev/snips/tract/target/debug/tract model.onnx -i 100x10xi8 --pass analyse dump`
┏ 0 Source input
┃   ━━━ 100x10xI8
┣ 11 AddDims input.unsqueeze
┃   ━━━ 1x100x10xI8
┣ 12 PermuteAxes input.transposed
┃   ━━━ 1x10x100xI8
┣┻┻ 13 QuantizeLinear tdnn.preaffineq7.output
┣┻┻┻┻┻┻┻┻ 14 Conv tdnn.tdnn
┃   ━━━ 1x5x98xI8
┣┻┻ 15 DequantizeLinear tdnn.tdnn.float
┃   ━━━ 1x5x98xF32
┣ 16 ScalarMax tdnn.relu
┣┻┻ 17 QuantizeLinear tdnn.output
┃   ━━━ 1x5x98xI8
┣┻┻ 18 DequantizeLinear tdnn.relu.q7.dequantize.output
┃   ━━━ 1x5x98xF32
┣ 19 PermuteAxes tdnn.relu.q7.dequantize.output.transposed
┃   ━━━ 1x98x5xF32
┣ 20 Squeeze output
    ━━━ 98x5xF32
```

**Comment by bminixhofer** [comment]
Thanks for your work on this!

But I'm a bit confused by the output above.
Are you referring to the model I linked above ([download from here](https://api.wandb.ai/files/bminixhofer/nnsplit/4phefspj/model/model.onnx))?

I'd expect the input to be of type `u8` not ì8` and there shouldn't be a Conv layer. When I visualize the model in [netron](https://lutzroeder.github.io/netron/) the only parametrized ops are `LSTM`and a `MatMul`and `Add` at the end. So the output should be `100x2`not `98x5`.

**Comment by kali** [comment]
(Car-pooling #270 here as the net uses bidi LSTM too)

**Comment by kali** [comment]
All right, I think I got the wrong model somehow. The one I'm look at now has two bidi lstm followed by matmul with bias. And is complaining about a missing cast that I will add promptly.

**Comment by kali** [comment]
And what look like more shape bugs after that. We'll get there.

**Comment by bminixhofer** [comment]
Exactly, that's the one. And I got the error with the cast (I believe it was u8 to i64) too. Great!

I'm not sure where that other model came from. I don't think it's from the W&B data (or at least I don't remember training one with that architecture). But funny that it lead to solving other bugs ;)

Thanks again for implementing this!

**Comment by kali** [comment]
your net hard codes the output shape, i had to disable it somewhere (this is a recurring issue in onnx, i'll have to do something about it).

But... look :)

```
TSAR 06/06 12:43 ~/dev/snips/tract/cli% RUST_BACKTRACE=1 cargo run -- bminixhofer.onnx -i 1x49xu8=0,0,0,0,0,68,97,115,32,105,115,116,32,101,105,110,32,84,101,115,116,32,68,97,115,32,105,115,116,32,110,111,99,104,32,101,105,110,32,84,101,115,116,46,0,0,0,0,0 --pass analyse run --dump
warning: unused variable: `fact`
   --> onnx/src/model.rs:154:51
    |
154 |             let pb::type_proto::Value::TensorType(fact) = fact;
    |                                                   ^^^^ help: if this is intentional, prefix it with an underscore: `_fact`
    |
    = note: `#[warn(unused_variables)]` on by default

warning: 1 warning emitted

    Finished dev [unoptimized + debuginfo] target(s) in 0.05s
     Running `/home/kali/dev/snips/tract/target/debug/tract bminixhofer.onnx -i 1x49xu8=0,0,0,0,0,68,97,115,32,105,115,116,32,101,105,110,32,84,101,115,116,32,68,97,115,32,105,115,116,32,110,111,99,104,32,101,105,110,32,84,101,115,116,46,0,0,0,0,0 --pass analyse run --dump`
output #0
1x49x2xF32 -14.299928, -13.624513, -18.213139, -16.244501, -17.66429, -14.819768, -15.950892, -13.127346, -12.452399, -9.47628, -16.920752, -12.54746, -12.663324, -10.356748, -16.663204, -14.129541, -10.692644, 15.813489, -13.061052, -10.428478, -14.32623, -13.423851, -17.565641, -12.581093, -9.206636, 15.954254, -10.393023, -9.971023, -11.869035, -12.145806, -16.26568, -12.524339, -5.8647366, 19.260675, -10.049599, -10.217356, -8.702155, -10.479693, -11.1579485, -9.932108, -13.198123, -6.3030777, 6.0882387, 20.08386, -16.230629, -14.483787, -10.280454, -10.167571, -14.925722, -13.585652, -8.539471, 15.871686, -12.033297, -10.155522, -13.952672, -13.346983, -16.931229, -13.072882, -10.52119, 15.736996, -12.926277, -11.761221, -13.520977, -12.525223, -13.31629, -11.495392, -15.235784, -13.022429, -7.9736633, 19.193443, -10.759927, -8.968889, -13.020223, -10.661898, -18.604511, -12.86927, -7.9958863, 18.667093, -13.915828, -12.533027, -12.08665, -12.4058275, -11.111715, -7.3786526, -13.389302, 10.80961, 14.302056, 17.48646, -10.359129, -12.230581, -13.6342125, -15.254908, -14.058313, -14.8817215, -15.836784, -15.133814, -12.929239, -10.459671
```

**Comment by bminixhofer** [comment]
Wow, that was quick!

Not 100% sure the output is correct but it sure looks like it :)
So only dynamic shapes are missing. I'm really curious to see how that performs in the browser now.

**Comment by kali** [comment]
Here we go, I already had a way to override a fact, it just could not be overriden by nothing...

``` cargo run -- bminixhofer.onnx -i 1x49xu8=0,0,0,0,0,68,97,115,32,105,115,116,32,101,105,110,32,84,101,115,116,32,68,97,115,32,105,115,116,32,110,111,99,104,32,101,105,110,32,84,101,115,116,46,0,0,0,0,0 --override-fact Add_26: run --dump```

**Comment by bminixhofer** [comment]
People seem to have recurring (pun intended) problems with that in PyTorch (https://github.com/onnx/onnx/issues/654). I'm not sure if tract should do something about it as it's inherently a problem with exporting.

I'll see if I can fix it in my ONNX export code.

**Comment by kali** [comment]
Well, it's not like it would the first problem that training frameworks generate that I have to deal with :) At least this one is easy to workaround from the CLI (and from the model API too).

**Comment by kali** [comment]
@bminixhofer, @bio43, is there a possibiity one of your networks with bidi lstm could be made part of the CI ? By that I mean 1/ you're ok with it, and  2/ they're un-encumbered by IP concerns, including the dataset you use to train.
In that case, I would of course need you to double check the results this branch is producing (comparing with onnx runtime is a way). I would be grateful if you could make them in the form of https://github.com/snipsco/tract/tree/master/onnx/test_cases/qrelu_1 (model.onnx, io.npz, with an empty vars.sh, and you can ignore the rest) that can be run by the script https://github.com/snipsco/tract/blob/master/onnx/test_cases/run_all.sh . io.npz internal names must match input and output nodes names so that the command line can make sense of it.

If the net is small enough, we'll just put it in git, if it's bigger, I'll put it in s3.

And then we can merge and release :)

**Comment by bminixhofer** [comment]
Sure, I can do that.

I'm completely fine with it. The code is open source under MIT and it was trained on German Wikipedia so there shouldn't be any issues.

**Comment by bio43** [comment]
Hi @kali , sure, my model can be used under MIT too. It is entirely untrained though (just has the initial weights that PyTorch assigned). I'm not sure how to generate the npz file, but the input is as described on the ticket (`3x1x24xf32` and `1xi32`) and the output (`3x1x5xf32` and `1xi32`), so should be fairly easy just to generate some random input data.

**Comment by kali** [comment]
@bio43 npz is just numpy savez_compressed format. It should be pretty natural if you're using PyTorch.

**Comment by bio43** [comment]
@kali yes, I looked up savez_compressed, but it seems to take a single matrix as input. What do I do when my model takes in multiple matrices like above? Also, how do I encode input/output?

**Comment by bio43** [comment]
(feel free to create the npz file yourself, since that might be faster than me trying to understand this :) )

**Comment by bminixhofer** [comment]
I tried this with a new model where I fixed the hardcoded dimension problem ([download here](https://api.wandb.ai/files/bminixhofer/nnsplit/3afwdumk/model/model.onnx)). 

But it doesn't work with batch sizes > 1. Here's the code to generate the input:

```python
import onnxruntime
import numpy as np

# list of bytes (including padding by zeros), sentence boundaries at byte 21 and 43
x = [0] * 5 + list("Das ist ein Test Das ist noch ein Test.".encode("utf-8")) + [0] * 5
x = np.array(x, dtype=np.float32)[None, :]  # shape 1 x len
x = np.concatenate([x, x], 0)  # shape 2 x len, REMOVE THIS LINE AND IT WORKS

sess = onnxruntime.InferenceSession("model.onnx")
input_name = sess.get_inputs()[0].name

# shape 1 x len x 2
# value 1 in last dimension are token boundaries, value 0 are sentence boundaries
# outputs are logits, in practice there would be a sigmoid afterwards
out = sess.run(None, {input_name: x})[0]

if len(out) > 1:
    assert np.allclose(
        out[0], out[1]
    )  # samples in batch are the same using onnxruntime

np.savez_compressed(open("io.npz", "wb"), input=x, Add_26=out)
```

When I run 
`cargo run -p tract -- model.onnx --input-bundle=io.npz run --dump --assert-output-bundle=io.npz`

 without the `concatenate` line I get the correct output. But when I run it with `concatenate` (i. e. batch size 2) I get an error. 

A warning I get on export in PyTorch could be relevant:

```
/Users/bminixhofer/Library/Python/3.7/lib/python/site-packages/torch/onnx/symbolic_opset9.py:1577: UserWarning: Exporting a model to ONNX with a batch_size other than 1, with a variable length with LSTM can cause an error when running the ONNX model with a different batch size. Make sure to save the model with a batch size of 1, or define the initial states (h0/c0) as inputs of the model. 
  "or define the initial states (h0/c0) as inputs of the model. ")
```

but I'm not sure what to do about it.

`onnxruntime` seems to handle this correctly as shown by the assert at the end of the Python code sample.

**Comment by kali** [comment]
Ok, there's a nasty bug on tract side.

**Comment by kali** [comment]
@bminixhofer Not sure it was clear, I have rebased the branch after the axis fixes, so it's ready for you to give it a shot again :)

**Comment by bminixhofer** [comment]
Thanks! I'll give it a go :)

**Comment by kali** [comment]
Thanks @bminixhofer ! We now have a test case that works on un-optimized mode and breaks on optim. I'll take it from here :)

**Comment by kali** [comment]
So, it's one more bug that comes from batch=2, and not the new bidi code, but I won't complain. ;)

**Comment by bminixhofer** [comment]
I'm happy that it works without optimize, I'm sure most of the work is done :)

BTW I have a TensorFlow equivalent of the model with the same architecture and same weights. I'm gonna use that to benchmark Tensorflow.js against tract once this branch is merged. That's gonna be really interesting.

**Comment by kali** [comment]
Here. Done.

**Comment by bminixhofer** [comment]
Awesome! Thanks for all your work on this!
